src/transform/src/redundant_join.rs

```diff
@@ -122,26 +131,37 @@ impl RedundantJoin {
                     // We descend into the values and the body with the naively
                     // extended context.
                     for value in values.iter_mut() {
-                        self.action(value, lets)?;
+                        self.action(value, ctx)?;
+                    }
+                    // Clear uses from the just visited recursive binding
+                    // definitions.
+                    for value in values.iter_mut() {
+                        ctx.remove_uses(value);
```

Comment:
There is a bug here (and I think in the `Let` case) that will cause double-counting when removing uses, and consequently premature removal of some references. I'll need to fix that before we can proceed with the review.